# Tutorial

The file [210427.campy_bordetella.ipynb](210427.campy_bordetella.ipynb) contains the `jupyter` notebook for the
*Campylobacter* and *Bordetella* homopolymer tract analyses for the manuscript
https://doi.org/10.1101/2021.06.02.446710. 

# Usage

the basic help can be invoked with `tatajuba -h`, which will look something like

```
tatajuba 1.0.2
Compare histograms of homopolymeric tract lengths, within context.
The complete syntax is:

 tatajuba  [-h|--help] [-v|--version] [-p|--paired] [-k|--kmer={2,...,32}] [-m|--minsize={1,...,32}] [-i|--minreads=<int>] [-d|--maxdist=<int>] [-l|--leven=<int>] [-t|--nthreads=<int>] -g|--gff=<genome.gff3|genome.gff3.gz> [--fasta=<genome.fna|genome.fna.gz>] <fastq files> [<fastq files>]... [-o|--outdir=<file>]

  -h, --help                       print a longer help and exit
  -v, --version                    print version and exit
  -p, --paired                     paired end (pairs of) files
  -k, --kmer={2,...,32}            kmer size flanking each side of homopolymer (default=25)
  -m, --minsize={1,...,32}         minimum homopolymer tract length to be compared (default=4)
  -i, --minreads=<int>             minimum number of reads for tract+context to be considered (default=5)
  -d, --maxdist=<int>              maximum distance between kmers of a flanking region to merge them into one context (default=1)
  -l, --leven=<int>                levenshtein distance between flanking regions to merge them into one context (after ref genome mapping)
  -t, --nthreads=<int>             suggested number of threads (default is to let system decide; I may not honour your suggestion btw)
  -g, --gff=<genome.gff3|genome.gff3.gz> reference genome file in GFF3, preferencially with sequence
  --fasta=<genome.fna|genome.fna.gz> reference genome file in fasta format, if absent from GFF3
  <fastq files>                    fastq file with reads (weirdly, fasta also possible as long as contains all reads and not only contigs)
  -o, --outdir=<file>              output directory, or 'random' for generating random dir name (default=current dir '.')
'Context' is the pair of flanking k-mers. If you have paired end files, then their order should be strictly f1_R1, f1_R2, f2_R1 etc.
  that is, R1 and R2 should be consecutive. A GFF3 reference genome must also be supplied, and bwa will create a series of index files
  if tatajuba can find the DNA sequences at end of GFF3 file, after pragma '##FASTA'

You can also supply a fasta file, e.g. when you download GFF3 from NCBI it may lack the DNA genome at the end --- then downloading the
  associated fna may work (untested). The internal library updates the GFF3 structure with provided fasta sequences (so a fasta file may
  overwrite DNA sequences present in the GFF3 file. If you modify the fasta file please delete the index files generated by the BWA library
  so that they can be regenerated with the updated information.

Notice that tatajuba creates files with same prefix and in same location as the GFF3 file, which may overwrite existing ones.
  On the other hand, as suggested above, you can recreate all the generated files by deleting them and running tatajuba again.

The default values are as following:
kmer= 25         minsize=  4     minread=  5     maxdist=  1
```

The minimum required information are the GFF3 file (assuming it has fasta sequences, as e.g. generated by `prokka`) and the reads. Usually the GFF3
file will not contain the sequence in fasta format, and thus you'll have to provide the fasta file with `--fasta` for
the exact same GFF3, with matching sequence names. 

Also, if you have paired reads you need to add option `-p` and make sure the read files are given in order, with reads
from same sample together: always `sampleA_1.fq sampleA_2.fq sampleB_1.fq sampleB_2.fq` but never
`sampleA_1.fq sampleB_1.fq sampleA_2.fq sampleB_2.fq`. 
Usually a wildcard (`*`) works since the reverse/forward flags are to the right of the sample IDs in the file names; to make sure you can try 
```
$ ls sample*
```
and see the order in which the files are shown.

## Mutation effect

Tatajuba can partition the homopolymer tracts (HTs) into those within or outside a coding region, and 
furthermore for those inside a coding region the change in tract length can gives us an idea about the effect on the
protein (multiples of three will lead to a missing amino acid, while other multiples will likely disrupt the downstream
coding region).

If you are interested in exploring the functional effect of mutations, there are a few tools available: 
[bcftools consequence calling](https://samtools.github.io/bcftools/howtos/csq-calling.html), 
[Ensembl Variant Effect Predictor (VEP)](https://www.ensembl.org/info/docs/tools/vep/index.html), and 
[snpEff](http://pcingola.github.io/SnpEff/).
They all rely on the VCF file for the samples, together with the GFF3 and FASTA files for the reference genome.

With exception of `bcftools csq`, which supports only ENSEMBL GFF3 files, we recommend `snpEff` or `VEP`. 
For `snpEff`, see instructions on using your own GFF3 file [from biostars](https://www.biostars.org/p/50963/) and from 
[its site](https://pcingola.github.io/SnpEff/se_buildingreg/#option-1-using-a-gff-file), where you'll have to build a database 
(directory with reference files). 
For VEP, instructions can be found [at the ENSEMBL site](https://www.ensembl.org/info/docs/tools/vep/script/vep_cache.html) &mdash; this software seems
to work better with GFF3 provided through [bp_genbank2gff3](https://metacpan.org/dist/BioPerl/view/bin/bp_genbank2gff3) (that is, converted from the full 
genbank file and removing the embedded FASTA entries at the end) than with GFF3 files downloaded directly from RefSeq...

The VCF files can be generated from the BAM/SAM files, i.e. using reference-based assembly programs like [minimap2](https://github.com/lh3/minimap2) or [bwa](https://github.com/lh3/bwa).
I particularly like [snippy](https://github.com/tseemann/snippy), which generates not only the BAM alignment and its corresponding `snps.filt.vcf` files, as it annotates the predicted effects with
`snpEff` into the `snps.vcf` files. Although the directories from the indivudal samples have relevant information,
please keep in mind that the "core SNPs" and [further analyses from snippy-multi](https://github.com/tseemann/snippy#core-snp-phylogeny) by design will exclude most information from the HTs.

Tatajuba will soon output a BED file with the genomic regions of interest. Tatajuba cannot be used as a variant caller
or aligner since (1) it does not use the full information available on the `fastq` files, focusing only on HTs and
neglecting many SNPs; (2) it processes and splits the reads before mapping to the reference genome &mdash; in particular 
one read segment can belong to several HTs (as part of their flanking region, for instance), and one HT can contain
variable segments (flanking regions _and_ HT length).


